pkg/apis/query/v1alpha1/cache_types.go

```diff
@@ -0,0 +1,151 @@
+	ResourceNamespaceScopedClusterCache = false
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Cluster"
+
+// ClusterCache defines a list of member cluster to be cached.
+type ClusterCache struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec represents the desired behavior of ClusterCache.
+	Spec ClusterCacheSpec `json:"spec,omitempty"`
+
+	// Status represents the status of ClusterCache.
+	// +optional
+	Status ClusterCacheStatus `json:"status,omitempty"`
+}
+
+// ClusterCacheSpec defines the desired state of ClusterCache
+type ClusterCacheSpec struct {
+	// ClusterSelectors represents the filter to select clusters.
+	// +required
+	ClusterSelectors []ClusterSelector `json:"clusterSelectors"`
+
+	// ResourceSelectors used to select resources.
+	// +required
+	ResourceSelectors []ResourceSelector `json:"resourceSelectors"`
+
+	// StatusUpdatePeriodSeconds is the period to update the status of the resource.
+	// default is 10s.
+	// +optional
+	StatusUpdatePeriodSeconds uint32 `json:"statusUpdatePeriodSeconds,omitempty"`
+}
+
+// ClusterSelector represents the filter to select clusters.
+type ClusterSelector struct {
+	// LabelSelector is a filter to select member clusters by labels.
+	// If non-nil and non-empty, only the clusters match this filter will be selected.
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// FieldSelector is a filter to select member clusters by fields.
+	// If non-nil and non-empty, only the clusters match this filter will be selected.
+	// +optional
+	FieldSelector *FieldSelector `json:"fieldSelector,omitempty"`
+
+	// ClusterNames is the list of clusters to be selected.
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// ExcludedClusters is the list of clusters to be ignored.
+	// +optional
+	ExcludeClusters []string `json:"exclude,omitempty"`
+}
+
+// FieldSelector is a field filter.
+type FieldSelector struct {
+	// A list of field selector requirements.
+	MatchExpressions []corev1.NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// ResourceSelector the resources will be selected.
+type ResourceSelector struct {
+	// APIVersion represents the API version of the target resources.
+	// +required
+	APIVersion string `json:"apiVersion"`
+
+	// Kind represents the Kind of the target resources.
+	// +required
+	Kind string `json:"kind"`
+
+	// Namespace of the target resource.
+	// Default is empty, which means all namespaces.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// ClusterCacheStatus defines the observed state of ClusterCache
+type ClusterCacheStatus struct {
+	// +optional
+	ResourceStatuses []ResourceStatusRef `json:"resources,omitempty"`
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
```

Comment:
I'm not so confident with the status filed, how about leaving it empty for now, just add an empty `Condition`?, like:
```golang
// ClusterCacheStatus defines the observed state of ClusterCache
type ClusterCacheStatus struct {
	// Conditions contain the different condition statuses.
	// +optional
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}
```

Reply:
Good suggestion

Reply:
Yeah, thanks. It's easy to add more fields to it, but not to remove(deprecate) them.